src/wp-includes/blocks.php

```diff
@@ -786,7 +791,68 @@ function get_hooked_block_markup( &$anchor_block, $hooked_block_type ) {
 	// However, its presence does not affect the frontend.
 	$anchor_block['attrs']['metadata']['ignoredHookedBlocks'][] = $hooked_block_type;
 
-	return get_comment_delimited_block_content( $hooked_block_type, array(), '' );
+	return serialize_block( $hooked_block );
+}
+
+/**
+ * Returns the markup for blocks hooked to the given anchor block in a specific relative position.
+ *
+ * @since 6.5.0
+ * @access private
+ *
+ * @param array                   $anchor_block      The anchor block.
+ * @param string                  $relative_position The relative position of the hooked blocks.
+ *                                                   Can be one of 'before', 'after', 'first_child', or 'last_child'.
+ * @param array                   $hooked_blocks     An array of hooked blocks, grouped by anchor block and relative position.
+ * @param WP_Block_Template|array $context           The block template, template part, or pattern that the anchor block belongs to.
+ * @return string
+ */
+function insert_hooked_blocks( &$anchor_block, $relative_position, $hooked_blocks, $context ) {
+	$anchor_block_type  = $anchor_block['blockName'];
+	$hooked_block_types = isset( $hooked_blocks[ $anchor_block_type ][ $relative_position ] )
+		? $hooked_blocks[ $anchor_block_type ][ $relative_position ]
+		: array();
+
+	/**
+	 * Filters the list of hooked block types for a given anchor block type and relative position.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string[]                $hooked_block_types The list of hooked block types.
+	 * @param string                  $relative_position  The relative position of the hooked blocks.
+	 *                                                    Can be one of 'before', 'after', 'first_child', or 'last_child'.
+	 * @param string                  $anchor_block_type  The anchor block type.
+	 * @param WP_Block_Template|array $context            The block template, template part, or pattern that the anchor block belongs to.
+	 */
+	$hooked_block_types = apply_filters( 'hooked_block_types', $hooked_block_types, $relative_position, $anchor_block_type, $context );
+
+	$markup = '';
+	foreach ( $hooked_block_types as $hooked_block_type ) {
+		$hooked_block = array(
```

Comment:
It's a very low-level concept as it's the shape of the parsed block. Are we sure that we want to open `innerBlocks` plus `innerContent`? What alternatives do we have here to represent the shape in something more abstract that doesn't impact performance?

Reply:
Yeah, that's a good question. I was originally only thinking to allow modifying attributes, e.g.

```php
$hooked_block_attributes = apply_filters( "hooked_block_attributes_{$hooked_block_type}", $hooked_block_attributes, $hooked_block_type, $relative_position, $anchor_block, $context );
```

(Note how `$hooked_block_attributes` is both the first attribute and the return value. I believe we'd need to do it like that (rather than making `$hooked_block_type` the first argument and return value), since we'd expect filters to change attributes, while the block type is generally going to remain the same.)

There were a number of reasons against this kind of filter signature; I'll try to list them by relevance:
1. Discussion over at #5811 revealed that extenders might still want to wrap their hooked block in a Group block, in order to get the `layout` attribute to work properly, as I was cautioned by folks more familiar with Layout block-support (in case it's impossible to add it directly to the hooked block). Additionally, we've heard of other use cases where people would like to inject e.g. two hooked blocks inside of a shared Row block parent. Allowing `innerBlocks`/`innerContent` to be set provides at least a low-level escape hatch for these use cases.
2. Only allowing attributes to be set would raise some questions about the filter signature, if we also want to provide access to the anchor block (which we do). Would we provide the entire parsed anchor block (even though it's more inconsistent with having only the hooked block type and attributes)? Or would we also only provide the anchor block's type and attributes (which is more consistent, but lacks information about inner blocks)?
3. We already have the parsed block format, and `WP_Block` class instances; I didn't want to add yet another way of specifying a block 😬 

Overall, I agree that the parsed block format isn't exactly pretty; I just found it rather hard to come up with a better format or function signature that worked better. Allowing access to `innerBlocks` and `innerContent` is kind of a side effect, but not totally unwanted (see 1.).

But if you do have an idea how we could do this differently, please LMK!

Reply:
What if we pass the result of `get_hooked_block_markup` call instead? In the majority of cases, that would never change, but if plugins would like to produce a different markup, then they can use the filter to rewrite the HTML to whatever they want:
- change the attributes
- wrap with a parent block
- rewrite it to any HTML they want, which is the only risky but the same issue exists with `innerBlocks` and `innerContent` as you can put there anything really

I don't think it's any different from what the API currently offers, but we operate on a much higher level without forcing folk to learn the parsed block structure.

Reply:
> What if we pass the result of `get_hooked_block_markup` call instead?

Hmm, that would be serialized block markup, i.e. something like `<!-- my/hooked-block /-->`.

> In the majority of cases, that would never change, but if plugins would like to produce a different markup, then they can use the filter to rewrite the HTML to whatever they want:
> 
> * change the attributes
> * wrap with a parent block
> * rewrite it to any HTML they want, which is the only risky but the same issue exists with `innerBlocks` and `innerContent` as you can put there anything really

But even for a simple operation like changing (or adding) an attribute, they would either have to do some string/RegEx-gymnastics (which can be error-prone if they get the markup wrong), or run `parse_blocks` again on something that was just parsed and re-serialized.

My thinking is that we want to make it easy enough to do a "simple" (and often-requested) change like set/change attributes (with the current state of this PR: `$hooked_block['attrs']['layout'] = $anchor_block['attrs']['layout'];`), and a bit harder to wrap with a parent or include inner blocks; it seems reasonable to require that an extender "knows what they're doing" if they want to do something more complex like that. (If they'd like to avoid manually nesting arrays to create the parsed block structure, they could still run `parse_blocks` on a string literal, e.g. `parse_blocks( <-- wp:group /--><div class="wp-block-group"><!-- my/hooked-block /--></div><-- /wp:group -->' )`.)

> I don't think it's any different from what the API currently offers, but we operate on a much higher level without forcing folk to learn the parsed block structure.

Personally, I see markup as lower-level than parsed block arrays 😅  but that might be subjective. It's true however that extenders might be familiar with (block) markup, but not necessarily with parsed block arrays.

FWIW, some of our filters in the render pipeline -- e.g. [`render_block`](https://developer.wordpress.org/reference/hooks/render_block/) or [`render_block_data`](https://developer.wordpress.org/reference/hooks/render_block_data/) -- also pass parsed block arrays as arguments; this was partly the inspiration here 🙂

Reply:
> FWIW, some of our filters in the render pipeline -- e.g. [render_block](https://developer.wordpress.org/reference/hooks/render_block/) or [render_block_data](https://developer.wordpress.org/reference/hooks/render_block_data/) -- also pass parsed block arrays as arguments; this was partly the inspiration here 🙂

Interesting observation. In effect, let's make it clear then and use `$parsed_hooked_block` for clarity.